### PR TITLE
Video streaming

### PR DESCRIPTION
**Pull request by verycumbersome**
#6 Peep this PR for any strange behavior or something weird. I extended a base template to allow for multiple pages for different cameras/masks, but since the camera changes depending on the network table value, I'm thinking the camerarear url should maybe just be used for the camera mask and the camera switching can be determined by the network table?

### CODE REVIEW

**Comment by andrewda** [comment]
Damn dude this is awesome!

> I'm thinking the camerarear url should maybe just be used for the camera mask and the camera switching can be determined by the network table?

Yea that probably makes sense, especially cause we probably won't have 2 cameras every year.

**Comment by verycumbersome** [comment]
Right I forgot that next years game might not have both cameras lol. Changed the camera from camera rear scheme to be camera mask and camera main

File: vision/app.py [hunk 1]
```diff
@@ -80,91 +82,107 @@ def run_image(self):
...
+			im = self.cameraRear.read()
+			try:
+				lowerThreshold = np.array([nt_utils.get_number("rear_lower_blue"), nt_utils.get_number("rear_lower_green"), nt_utils.get_number("rear_lower_red")])
+				upperThreshold = np.array([nt_utils.get_number("rear_upper_blue"), nt_utils.get_number("rear_upper_green"), nt_utils.get_number("rear_upper_red")])
+			except:
+				lowerThreshold = self.lower
+				upperThreshold = self.upper
+
+		if im is not None:
+			im = cv2.resize(im, (600, 480), 0, 0)
```
Comment (iansabine) [inline vision/app.py]: I'm sure there's a good reason for this number to be 600, but most camera's aspect ratio would be 4:3 which would make the size 640 x 480.

Comment (andrewda) [reply vision/app.py]: That's a good point. I think we might have used this last year accidentally for some reason.

Comment (iansabine) [reply vision/app.py]: Also, use constants instead of these magic numbers.

**Comment by andrewda** [comment]
Failing test is irrelevant and fixed in #11

**Comment by andrewda** [comment]
This works really well! Again, thanks a million Matthew! I'm gonna close it for now, though, as we're going to probably want to redo the entire logic a little bit before getting something like this in. I want streaming to be more of a high-level feature than having it run its own instance of the app.

At the moment, it can only be accessed from one computer or even only one session at a time – ideally it would be completely independent of the `Vision().run()` function, and even `Vision().get_frame()`. I've got several idea regarding this, including maybe even just saving a `t.jpg` in the main process if `--stream` is set and using that (though blocking fs operations might be slower than we want). Another (probably cleaner) option would be to have a getter in `Vision` to access the latest frame, and read that in `stream.py`.

All in all, this is fantastic and we'll definitely be using most of this for what ends up being the streaming code, but there needs to be some fundamental changes to the app before it's doable.

**Pull request closed without merging**